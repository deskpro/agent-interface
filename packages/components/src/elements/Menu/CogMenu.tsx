import * as React from "react";
import classNames from "classnames";
import { Manager, Reference, Popper } from "react-popper";

import Icon from "../Icon/Icon";
import { MenuProps } from "./Menu";
import useOutsideClick from "../../hooks/useOutsideClick";
import useMenu from "../../hooks/useMenu";

export type CogProps = {
  renderMenu: (menuProps: Partial<MenuProps>) => React.ReactElement<any>;
  className?: string;
  isVisible?: boolean;
  style?: any;
  placement?: string;
  innerRef?: React.Ref<HTMLSpanElement>;
};

const Cog: React.FC<CogProps> = ({
  className,
  renderMenu,
  isVisible = false,
  style,
  placement,
  innerRef
}) => {
  const { isVisible: menuIsVisible, hideMenu, toggleMenu } = useMenu();

  const cogRef = React.useRef<HTMLSpanElement>(null);
  useOutsideClick(cogRef, hideMenu);

  const [isMouseOver, setIsMouseOver] = React.useState(false);

  if (!isVisible && !menuIsVisible && !isMouseOver) {
    return null;
  }

  return (
    <span
      className={classNames("dp-Cog", className, {
        "is-active": menuIsVisible
      })}
      style={style}
      role="button"
      tabIndex={-1}
      ref={ref => {
        (cogRef.current as HTMLSpanElement | null) = ref;
        if (typeof innerRef === "function") {
          innerRef(ref);
        } else if (innerRef && innerRef.current) {
          // eslint-disable-next-line no-param-reassign
          (innerRef.current as HTMLSpanElement | null) = cogRef.current;
        }
      }}
      data-placement={placement}
      onMouseEnter={() => setIsMouseOver(true)}
      onMouseLeave={() => {
        if (!menuIsVisible) {
          setIsMouseOver(false);
        }
      }}
    >
      <Manager>
        <Reference>
          {({ ref }) => (
            <Icon
              name="settings"
              size={12}
              circle
              onClick={toggleMenu}
              ref={ref}
            />
          )}
        </Reference>
        {menuIsVisible && (
          <Popper
            placement="bottom-start"
            modifiers={{
              preventOverflow: {
                enabled: true,
                escapeWithReference: true,
                boundariesElement: "viewport"
              },
              flip: {
                enabled: true,
                flipVariationsByContent: true
              }
            }}
          >
            {({ ref, style: menuStyles }) =>
              renderMenu({
                menuRef: ref,
                style: menuStyles,
                onMenuClose: hideMenu
              })
            }
          </Popper>
        )}
      </Manager>
    </span>
  );
};

export default React.memo(
  React.forwardRef<HTMLSpanElement, CogProps>((props, ref) => (
    <Cog {...props} innerRef={ref} />
  ))
);
