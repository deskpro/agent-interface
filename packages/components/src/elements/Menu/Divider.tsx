import * as React from "react";

const MenuDivider: React.FC = () => <li className="dp-DropdownDivider" />;

export default MenuDivider;
