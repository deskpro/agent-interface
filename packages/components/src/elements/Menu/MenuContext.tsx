import * as React from "react";
import produce from "immer";

type Position = { x: number; y: number };

export type MenuContextType = {
  visibleMenu: string | null;
  position: Position;
  show: (
    e: React.SyntheticEvent,
    args: {
      menuId: string;
      menuType?: "context" | "dropdown" | "cog" | "simple";
    }
  ) => void;
  hide: () => void;
  updatePosition: (position: Position) => void;
};

export const MenuContext = React.createContext<MenuContextType>({
  visibleMenu: null,
  position: { x: 0, y: 0 },
  show: () => {},
  hide: () => {},
  updatePosition: () => {}
});

/* eslint-disable no-param-reassign */
const menuReducer = produce((state: MenuContextType, { type, payload }) => {
  switch (type) {
    case "show": {
      state.visibleMenu = payload.menuId;
      const { e } = payload;
      let pos;

      if (payload.menuType === "context") {
        pos = {
          x: e.clientX,
          y: e.clientY
        };

        if (
          e.type === "touchend" &&
          (!pos.x || !pos.y) &&
          (e.changedTouches && e.changedTouches.length > 0)
        ) {
          pos.x = e.changedTouches[0].clientX;
          pos.y = e.changedTouches[0].clientY;
        }

        if (!pos.x || pos.x < 0) {
          pos.x = 0;
        }

        if (!pos.y || pos.y < 0) {
          pos.y = 0;
        }
      } else if (payload.menuType === "cog") {
        pos = {
          x: 2,
          y: 5
        };
      } else if (payload.menuType === "dropdown") {
        pos = {
          x: e.target.offsetLeft,
          y: e.target.offsetTop + e.target.offsetHeight + 3
        };
      }

      state.position = pos;
      break;
    }

    case "hide":
      state.visibleMenu = null;
      break;

    case "updatePosition":
      state.position = payload;
      break;

    default:
      break;
  }
});
/* eslint-enable no-param-reassign */

export const ContextMenuProvider: React.FC = ({ children }) => {
  const [state, dispatch] = React.useReducer(menuReducer, {
    visibleMenu: null,
    position: { x: 0, y: 0 },
    show: (e: React.SyntheticEvent, args) => {
      if (e.persist) {
        e.persist();
      }
      dispatch({ type: "show", payload: { e, ...args } });
    },
    hide: () => {
      dispatch({ type: "hide" });
    },
    updatePosition: position => {
      dispatch({ type: "updatePosition", payload: position });
    }
  });

  return <MenuContext.Provider value={state}>{children}</MenuContext.Provider>;
};
