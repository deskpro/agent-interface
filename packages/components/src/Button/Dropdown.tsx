import * as React from "react";
import { Manager, Reference, Popper } from "react-popper";
import classNames from "classnames";

import "@deskpro/agent-interface-style/dist/dp-Arrow.css";
import { MenuProps } from "../elements/Menu/Menu";
import useOutsideClick from "../hooks/useOutsideClick";

export type DropdownProps = {
  className?: string;
  renderMenu: (menuProps: Partial<MenuProps>) => React.ReactElement<any>;
};

const Dropdown: React.FC<DropdownProps> = ({
  className,
  children,
  renderMenu
}) => {
  const [isOpen, setOpenState] = React.useState(false);
  const dropdownRef = React.useRef<HTMLSpanElement>(null);
  useOutsideClick(dropdownRef, () => {
    setOpenState(false);
  });

  return (
    <span className="dp-Dropdown-container" ref={dropdownRef}>
      <Manager>
        <Reference>
          {({ ref }) => (
            <span
              className={classNames("dp-Dropdown", className)}
              ref={ref}
              role="button"
              tabIndex={-1}
              onClick={() => setOpenState(!isOpen)}
              onKeyPress={e => {
                if (e.key === "Enter") setOpenState(!isOpen);
              }}
            >
              {children}
              <span className="dp-Arrow" />
            </span>
          )}
        </Reference>
        {isOpen && (
          <Popper
            placement="bottom-start"
            modifiers={{
              preventOverflow: {
                enabled: true,
                escapeWithReference: true,
                boundariesElement: "viewport"
              },
              flip: {
                enabled: true,
                flipVariationsByContent: true
              }
            }}
          >
            {({ ref, style }) =>
              renderMenu({
                menuRef: ref,
                style,
                onMenuClose: () => setOpenState(false)
              })
            }
          </Popper>
        )}
      </Manager>
    </span>
  );
};

export default Dropdown;
