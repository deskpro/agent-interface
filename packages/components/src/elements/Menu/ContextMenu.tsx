import * as React from "react";
import { Manager, Popper } from "react-popper";

import { MenuProps } from "./Menu";

import useOutsideClick from "../../hooks/useOutsideClick";
import useMenu from "../../hooks/useMenu";

type BoundingRect = {
  top: number;
  bottom: number;
  left: number;
  right: number;
  width: number;
  height: number;
};
interface VirtualReferenceElementType {
  getBoundingClientRect: () => BoundingRect;
  clientWidth: number;
  clientHeight: number;
}

export type ContextMenuProps = {
  triggerEvent?: "onClick" | "onContextMenu" | "onDoubleClick";
  renderMenu: (menuProps: Partial<MenuProps>) => React.ReactElement<any>;
};

const ContextMenu: React.FC<ContextMenuProps> = ({
  triggerEvent = "onContextMenu",
  renderMenu,
  children,
  ...props
}) => {
  const { isVisible: menuIsVisible, hideMenu, showMenu } = useMenu();

  const menuRef = React.useRef<HTMLUListElement>(null);
  useOutsideClick(menuRef, hideMenu);

  const virtualReference = React.useRef<VirtualReferenceElementType>(null);
  const handleContextMenuClick = React.useCallback(e => {
    e.preventDefault();

    const pos = {
      left: e.clientX,
      right: e.clientX,
      top: e.clientY,
      bottom: e.clientY,
      width: 1,
      height: 1
    };

    if (
      e.type === "touchend" &&
      (!pos.left || !pos.top) &&
      (e.changedTouches && e.changedTouches.length > 0)
    ) {
      pos.left = e.changedTouches[0].clientX;
      pos.top = e.changedTouches[0].clientY;
    }

    if (!pos.left || pos.left < 0) {
      pos.left = 0;
    }

    if (!pos.top || pos.top < 0) {
      pos.top = 0;
    }

    pos.right = pos.left + pos.width;
    pos.bottom = pos.top + pos.height;

    (virtualReference.current as VirtualReferenceElementType) = {
      getBoundingClientRect() {
        return pos;
      },
      get clientWidth() {
        return pos.width;
      },
      get clientHeight() {
        return pos.height;
      }
    };

    showMenu();
  }, []);

  return (
    <Manager>
      <span
        className="dp-ContextMenuWrapper"
        style={{ display: "inline-block" }}
        {...{
          [triggerEvent]: handleContextMenuClick
        }}
        {...props}
      >
        {children}
      </span>
      {menuIsVisible && (
        <Popper
          referenceElement={
            virtualReference.current as VirtualReferenceElementType
          }
          placement="bottom-start"
          modifiers={{
            preventOverflow: {
              enabled: true,
              escapeWithReference: true,
              boundariesElement: "viewport"
            },
            flip: {
              enabled: true,
              flipVariationsByContent: true
            }
          }}
        >
          {({ ref, style }) =>
            renderMenu({
              style,
              menuRef: (el: HTMLUListElement) => {
                ref(el);
                (menuRef.current as HTMLUListElement) = el;
              },
              onMenuClose: hideMenu
            })
          }
        </Popper>
      )}
    </Manager>
  );
};

export default ContextMenu;
