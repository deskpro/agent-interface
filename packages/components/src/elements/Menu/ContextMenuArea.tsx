import * as React from "react";
import { MenuContext } from "./MenuContext";

export type ContextMenuAreaProps = {
  menuId: string;
  triggerEvent?: "onClick" | "onContextMenu" | "onDoubleClick";
};

const ContextMenuArea: React.FC<ContextMenuAreaProps> = ({
  triggerEvent = "onContextMenu",
  menuId,
  ...props
}) => {
  const context = React.useContext(MenuContext);

  return (
    <div
      style={{ display: "inline-block" }}
      {...{
        [triggerEvent]: (e: React.SyntheticEvent) => {
          e.preventDefault();
          context.show(e, { menuId, menuType: "context" });
        }
      }}
      {...props}
    />
  );
};

export default ContextMenuArea;
