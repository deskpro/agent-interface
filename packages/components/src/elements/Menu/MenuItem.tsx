import * as React from "react";
import classNames from "classnames";
import { Manager, Reference, Popper } from "react-popper";

import Icon from "../Icon/Icon";
import { MenuContext } from "./MenuContext";

export type MenuItemProps = {
  icon?: string | React.ReactElement<any>;
  name: React.Key;
  text: string;
  link?: string;
  onClick?: (itemName: React.Key, e: React.SyntheticEvent<HTMLElement>) => void;
  isChecked?: boolean;
  className?: string;
  linkComponent?: any;
  trail?: string;
};

const MenuItem: React.FC<MenuItemProps> = ({
  text,
  name,
  icon,
  link,
  onClick,
  className,
  children,
  isChecked = false,
  linkComponent = "a",
  trail = "",
  ...itemProps
}) => {
  const { activeTrail, setActiveTrail } = React.useContext(MenuContext);
  const clickHandler = React.useCallback(
    e => {
      if (typeof onClick === "function") {
        onClick(name, e);
      }
    },
    [name, onClick]
  );
  const activateMenuItem = React.useCallback(
    () => {
      setActiveTrail(trail);
    },
    [trail, setActiveTrail]
  );

  const labelProps = {
    className: "dp-Menu-link"
  };
  if (link) {
    labelProps[linkComponent === "a" ? "href" : "to"] = link;
  }

  return (
    <li
      className={classNames("dp-Menu-linkItem", className, {
        "is-checked": isChecked
      })}
      {...itemProps}
      role="menuitem"
      onClick={clickHandler}
      onKeyPress={clickHandler}
      onMouseOver={activateMenuItem}
      onFocus={activateMenuItem}
    >
      <Manager>
        <Reference>
          {({ ref }) => (
            <span ref={ref} style={{ display: "inline-block", width: "100%" }}>
              {!!icon && typeof icon === "string" && (
                <Icon name={icon} size={13} />
              )}
              {!!icon && typeof icon !== "string" && icon}
              {React.createElement(
                link ? linkComponent : "span",
                labelProps,
                text
              )}
              {!!children && <span className="dp-Arrow" />}
            </span>
          )}
        </Reference>
        {!!children && activeTrail.startsWith(trail) && (
          <Popper
            placement="right-start"
            modifiers={{
              // offset: { enabled: true, offset: "0,11" },
              flip: {
                enabled: true,
                behavior: ["right", "left"],
                padding: 0,
                flipVariationsByContent: true
              },
              preventOverflow: {
                enabled: true,
                escapeWithReference: true,
                boundariesElement: "viewport"
              }
            }}
          >
            {({ ref, style, placement }) =>
              React.cloneElement(
                React.Children.only(children) as React.ReactElement<any>,
                {
                  menuRef: ref,
                  style,
                  className: `is-${placement}`,
                  parentTrail: trail
                }
              )
            }
          </Popper>
        )}
      </Manager>
    </li>
  );
};

export default React.memo<React.FC<MenuItemProps>>(MenuItem);
