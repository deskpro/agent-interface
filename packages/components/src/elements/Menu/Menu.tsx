import * as React from "react";
import classNames from "classnames";

import "@deskpro/agent-interface-style/dist/components/dp-menus.css";
import MenuItem from "./MenuItem";
import MenuSearchItem from "./MenuSearch";

type MenuSubComponents = {
  MenuItem: typeof MenuItem;
  MenuSearch: typeof MenuSearchItem;
};

export type MenuProps = {
  className?: string;
  title?: string;
  isVisible?: boolean;
  linkComponent?: any;
  children: React.ReactElement<any>[];
};

const hasIcons = (children: React.ReactNodeArray) =>
  (React.Children.toArray(children) as React.ReactElement<any>[]).some(
    (child: React.ReactElement<any>) => !!child.props.icon
  );

const Menu: React.FC<MenuProps> & MenuSubComponents = ({
  className,
  title,
  children,
  linkComponent = "a",
  isVisible = false,
  ...otherProps
}) => (
  <ul
    className={classNames("dp-Menu", className, {
      [`Menu--icons`]: hasIcons(children),
      "is-visible": isVisible
    })}
    {...otherProps}
  >
    {!!title && <li className="dp-Menu-title">{title}</li>}
    {React.Children.map(
      children as React.ReactElement<any>[],
      (child: React.ReactElement<any>) =>
        React.cloneElement(child, {
          linkComponent
        })
    )}
  </ul>
);

Menu.MenuItem = MenuItem;
Menu.MenuSearch = MenuSearchItem;

export default Menu;
