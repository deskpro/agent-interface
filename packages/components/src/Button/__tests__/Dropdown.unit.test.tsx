import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import Dropdown from "../Dropdown";
import Menu from "../../elements/Menu/Menu";

describe("<Dropdown />", () => {
  it("should match its snapshot", () => {
    const { getByText, container } = render(
      <Dropdown
        renderMenu={menuProps => (
          <Menu {...menuProps}>
            <Menu.MenuItem
              name="select-all"
              text="Select All"
              onClick={jest.fn()}
            />
            <Menu.MenuItem
              name="select-none"
              text="Select none"
              onClick={jest.fn()}
            />
            <Menu.MenuItem name="inverse" text="Inverse" onClick={jest.fn()} />
          </Menu>
        )}
      >
        Dropdown button
      </Dropdown>
    );
    expect(getByText("Dropdown button")).toMatchSnapshot();
    expect(container.querySelector(".dp-Menu-link")).toBeNull();
  });
  it("should open on click", () => {
    const { getByText, container } = render(
      <div>
        <Dropdown
          renderMenu={menuProps => (
            <Menu {...menuProps}>
              <Menu.MenuItem
                name="select-all"
                text="Select All"
                onClick={jest.fn()}
              />
              <Menu.MenuItem
                name="select-none"
                text="Select none"
                onClick={jest.fn()}
              />
              <Menu.MenuItem
                name="inverse"
                text="Inverse"
                onClick={jest.fn()}
              />
            </Menu>
          )}
        >
          Dropdown button
        </Dropdown>
        <div>Outside box</div>
      </div>
    );

    expect(container.querySelector(".dp-Menu-link")).toBeNull();

    fireEvent(
      getByText("Dropdown button"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true
      })
    );
    expect(getByText("Select All")).toMatchSnapshot();
    fireEvent(
      getByText("Outside box"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true
      })
    );
  });
});
