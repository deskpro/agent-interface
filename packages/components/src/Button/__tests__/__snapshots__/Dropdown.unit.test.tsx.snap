// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should match its snapshot 1`] = `
<span
  class="dp-Dropdown"
  role="button"
  tabindex="-1"
>
  Dropdown button
  <span
    class="dp-Icon dp-Arrow"
    role="button"
    tabindex="-1"
  />
</span>
`;

exports[`<Dropdown /> should open on click 1`] = `
<span
  class="dp-Menu-link"
>
  Select All
</span>
`;
